lib/ansible/galaxy/dependency_resolution/dataclasses.py

```diff
@@ -428,11 +430,13 @@ def from_requirement_dict(cls, collection_req, art_mgr, validate_signature_optio
         if req_type not in {'galaxy', 'subdirs'} and req_version == '*':
             req_version = art_mgr.get_direct_collection_version(tmp_inst_req)
 
-        return cls(
+        req = cls(
             req_name, req_version,
             req_source, req_type,
             req_signature_sources,
         )
+        req.assert_version()
```

Comment:
Making assertions when an object holding invalid data is already initialized is an anti-pattern. This should be checked in or before init...